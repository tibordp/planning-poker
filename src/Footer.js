/**
 * MIT License
 *
 * Copyright (c) 2020 Tibor Djurica Potpara
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React from "react";
import PropTypes from "prop-types";
import GitHubIcon from "@material-ui/icons/GitHub";
import { makeStyles } from "@material-ui/core/styles";
import Box from "@material-ui/core/Box";
import Divider from "@material-ui/core/Divider";
import Button from "@material-ui/core/Button";
import { Timer } from "./Timer";
import { useSnackbar } from "notistack";
import { usePermissions } from "./permissions";
import { useRouter } from "next/router";
import { useConfirmationDialog } from "./utils/useConfirmationDialog";

const useStyles = makeStyles((theme) => ({
  button: {
    color: theme.palette.grey[700],
  },
  footer: {
    display: "flex",
    flexDirection: "row",
    flexWrap: "wrap",
    alignItems: "center",
    justifyContent: "space-between",
    minHeight: theme.spacing(6),
    "& > *:only-child": {
      marginLeft: "auto",
    },
  },
  timer: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
  },
  dangerButton: {
    color: theme.palette.error.main,
    borderColor: theme.palette.error.main,
  },
}));

export default function Footer({ sessionName, canReactivate, remoteState, dispatch }) {
  const classes = useStyles();
  const { enqueueSnackbar } = useSnackbar();
  const router = useRouter();

  const showTimer = !!remoteState && remoteState.settings.showTimer;
  const permissions = usePermissions(remoteState);

  const [confirmFinish, dialog] = useConfirmationDialog({
    title: "Finish this session?",
    description:
      "Finishing the session will disconnect everyone and afterwards you will be able to review the report.",
    confirmText: "Finish session",
    onConfirm: () => dispatch?.({ action: "finishSession" }),
  });

  const reactivateSession = async () => {
    try {
      const res = await fetch(`/api/sessions/${encodeURIComponent(sessionName)}/reactivate`, {
        method: "POST",
      });
      if (!res.ok) {
        throw new Error(res.statusText);
      }
      router.replace({
        pathname: "/[session]",
        query: { session: sessionName },
      });
    } catch (e) {
      console.error(e);
      enqueueSnackbar("Failed to reactivate the session!", { variant: "error" });
    }
  };

  return (
    <>
      <Divider />
      <Box className={classes.footer}>
        {showTimer && (
          <>
            <Box className={classes.footer}>
              <Timer
                canControlTimer={permissions.canControlTimer}
                timerState={remoteState.timerState}
                dispatch={dispatch}
              />
            </Box>
          </>
        )}
        {dispatch && permissions.canFinishSession && (
          <Button className={classes.button} variant="text" onClick={confirmFinish}>
            Finish session
          </Button>
        )}
        {canReactivate && sessionName && (
          <Button className={classes.button} variant="text" onClick={reactivateSession}>
            Reactivate session
          </Button>
        )}
        <Button
          variant="text"
          target="_blank"
          href="http://www.github.com/tibordp/planning-poker"
          color="default"
          className={classes.button}
          startIcon={<GitHubIcon />}
        >
          GitHub
        </Button>
      </Box>
      {dialog}
    </>
  );
}

Footer.propTypes = {
  remoteState: PropTypes.object,
  dispatch: PropTypes.func,
  canReactivate: PropTypes.bool,
  sessionName: PropTypes.string,
};
