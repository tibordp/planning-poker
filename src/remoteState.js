/**
 * MIT License
 *
 * Copyright (c) 2020 Tibor Djurica Potpara
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React from "react";
import { HeartbeatingWebsocket } from "./websocket";

export const IS_SSR = typeof navigator === "undefined" || typeof window === "undefined";

export function useInternetConnectivity() {
  const [haveConnectivity, setHaveConnectivity] = React.useState(IS_SSR || navigator.onLine);
  const setConnectivity = () => setHaveConnectivity(navigator.onLine);

  React.useEffect(() => {
    if (IS_SSR) {
      return () => {};
    }

    window.addEventListener("online", setConnectivity);
    window.addEventListener("offline", setConnectivity);
    setConnectivity();

    return () => {
      window.removeEventListener("online", setConnectivity);
      window.removeEventListener("offline", setConnectivity);
    };
  });

  return haveConnectivity;
}

export function useRemoteState(session) {
  const [dispatch, setDispatch] = React.useState(null);
  const [remoteState, setRemoteState] = React.useState(null);

  const haveConnectivity = useInternetConnectivity();

  React.useEffect(() => {
    if (IS_SSR || !haveConnectivity) {
      // If we are in SSR context or offline, do not try to continuously reconnect
      return () => {};
    }

    const protocol = location.protocol.replace("http", "ws");
    const host = `${location.hostname}:${location.port}`;
    const socket = new HeartbeatingWebsocket(`${protocol}//${host}/${session}`);

    socket.onopen = () => {
      setDispatch(() => (message) => socket.send(message));
    };

    socket.onmessage = (message) => {
      switch (message.action) {
        case "updateState":
          setRemoteState(message.value);
          break;
      }
    };
    socket.onclose = () => {
      setRemoteState(null);
      setDispatch(null);
    };

    return () => {
      socket.close();
    };
  }, [session, haveConnectivity]);

  return [remoteState, dispatch];
}

export const connectionState = {
  OFFLINE: "offline",
  CONNECTING: "connecting",
  RECONNECTING: "reconnecting",
  CONNECTED: "connected",
};

/**
 * Stores part of the state in sessionStorage (vote for the current epoch and
 * whether we are participant or not (along with name), so we can refresh the
 * page and stay connected seamlessly.
 */
export function useReconnector(session) {
  const [remoteState, dispatch] = useRemoteState(session);
  const haveConnectivity = useInternetConnectivity();

  React.useEffect(() => {
    if (!IS_SSR) {
      if (remoteState?.me) {
        window.sessionStorage.setItem(
          session,
          JSON.stringify({
            epoch: remoteState.epoch,
            settings: remoteState.settings,
            ...remoteState.me,
          })
        );
      }
    }
  }, [remoteState]);

  React.useEffect(() => {
    if (!IS_SSR) {
      const storedSession = window.sessionStorage.getItem(session);
      if (dispatch && storedSession) {
        dispatch({ action: "reconnect", ...JSON.parse(storedSession) });
      }
    }
  }, [dispatch]);

  let state;
  if (remoteState) {
    state = connectionState.CONNECTED;
  } else if (!haveConnectivity) {
    state = connectionState.OFFLINE;
  } else if (!IS_SSR && window.sessionStorage.getItem(session)) {
    state = connectionState.RECONNECTING;
  } else {
    state = connectionState.CONNECTING;
  }

  return [remoteState, dispatch, state];
}
