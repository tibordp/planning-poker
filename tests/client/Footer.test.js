import React from "react";
import renderer from "react-test-renderer";
import Footer from "../../src/Footer";
import { ThemeProvider } from "@material-ui/core";
import theme from "../../src/theme";
import { config } from "react-transition-group";
import * as remoteState from "./remoteState";

test.each([
  remoteState.remoteStateAsHost,
  remoteState.remoteStateAsVoterNoControl,
  remoteState.remoteStateNoTimer,
])("renders Footer unchanged %p", (state) => {
  config.disabled = true;
  const constantDate = new Date("2020-06-23T22:31:09.727Z");
  global.Date = class extends Date {
    constructor(...args) {
      super(...args);
      if (!args.length) {
        return constantDate;
      }
    }
  };

  const tree = renderer
    .create(
      <ThemeProvider theme={theme}>
        <Footer remoteState={state} dispatch={jest.fn()} />
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
