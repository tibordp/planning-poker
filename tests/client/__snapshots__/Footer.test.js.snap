// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Footer unchanged {"clients": [Array], "description": "### Markdown", "epoch": 0, "host": "5cdf4eb9-4c88-47ee-85da-5a2e970ed79f", "me": [Object], "pagination": [Object], "settings": [Object], "timerState": [Object], "votesVisible": false} 1`] = `
Array [
  <hr
    className="MuiDivider-root"
  />,
  <div
    className="MuiBox-root MuiBox-root-3 makeStyles-footer-2"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      00:00:00
    </p>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
          />
        </svg>
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
          />
        </svg>
      </span>
    </button>
    <div
      style={
        Object {
          "flexGrow": 1,
        }
      }
    />
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-1"
      href="http://www.github.com/tibordp/planning-poker"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
            />
          </svg>
        </span>
        View on GitHub
      </span>
    </a>
  </div>,
]
`;

exports[`renders Footer unchanged {"clients": [Array], "description": "### Markdown", "epoch": 0, "host": "5cdf4eb9-4c88-47ee-85da-5a2e970ed79f", "me": [Object], "pagination": [Object], "settings": [Object], "timerState": [Object], "votesVisible": false} 2`] = `
Array [
  <hr
    className="MuiDivider-root"
  />,
  <div
    className="MuiBox-root MuiBox-root-4 makeStyles-footer-2"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      01:00:00
    </p>
    <div
      style={
        Object {
          "flexGrow": 1,
        }
      }
    />
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-1"
      href="http://www.github.com/tibordp/planning-poker"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
            />
          </svg>
        </span>
        View on GitHub
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </a>
  </div>,
]
`;

exports[`renders Footer unchanged {"clients": [Array], "description": "### Markdown", "epoch": 0, "host": "5cdf4eb9-4c88-47ee-85da-5a2e970ed79f", "me": [Object], "pagination": [Object], "settings": [Object], "timerState": [Object], "votesVisible": false} 3`] = `
Array [
  <hr
    className="MuiDivider-root"
  />,
  <div
    className="MuiBox-root MuiBox-root-5 makeStyles-footer-2"
  >
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-1"
      href="http://www.github.com/tibordp/planning-poker"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      target="_blank"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
            />
          </svg>
        </span>
        View on GitHub
      </span>
    </a>
  </div>,
]
`;
